goblet/app.py

```diff
@@ -96,20 +103,75 @@ def deploy(
         if not skip_backend:
             log.info(f"preparing to deploy with backend {self.backend.resource_type}")
             source = backend.deploy(force=force)
-        if not skip_resources:
+
+        registered_handlers = self.get_registered_handler_resource_types()
+
+        # checks registered deployable handlers before determining if backend exists
+        if (
+            registered_handlers
+            and skip_backend
+            and (handlers or not skip_handlers)
+            and not backend.get()
+        ):
+            log.error("backend is not deployed, handlers cannot be deployed. exiting.")
+            sys.exit(1)
+
+        if handlers:
+            log.info(f"deploying {handlers} handlers")
+            self.deploy_handlers(source, handlers)
+        elif not skip_handlers:
+            log.info("deploying handlers")
             self.deploy_handlers(source)
 
-    def destroy(self, all=False, skip_infra=False):
+    def destroy(
+        self,
+        all=False,
+        skip_infra=False,
+        skip_handlers=False,
+        skip_backend=False,
+        handlers: List[str] = None,
+        infras: List[str] = None,
+    ):
         """Destroys http cloudfunction and then calls goblet.destroy() to remove handler's infrastructure"""
-        for k, v in self.handlers.items():
-            log.info(f"destroying {k}")
-            v.destroy()
-        self.backend.destroy(all=all)
-
-        if not skip_infra:
-            for k, v in self.infrastructure.items():
-                log.info(f"destroying {k}")
-                v.destroy()
+
+        if handlers:
+            log.info(f"destroying {handlers} handlers")
+            self.destroy_handlers(handlers)
+        elif not skip_handlers:
+            log.info("destroying handlers")
+            self.destroy_handlers()
+
+        if not skip_backend:
+            self.backend.destroy(all=all)
+
+        if infras:
+            log.info(f"destroying {infras} infrastructure")
+            self.destroy_infrastructure(infras)
+        elif not skip_infra:
+            log.info("destroying infrastructure")
+            self.destroy_infrastructure()
+
+    def sync(
+        self,
+        dryrun=False,
+        skip_infra=False,
+        skip_handlers=False,
+        handlers: List[str] = None,
+        infras: List[str] = None,
+    ):
+        if infras:
+            log.info(f"syncing {infras} infrastructure")
+            self.sync_infrastructure(dryrun, infras)
```

Comment:
should we add this logic to `sync_infrastructure` so that we dont have to call `sync_infrastructure` and `sync_handlers` seperately in the cli?

Reply:
the syncing functions need to be called separately, having one cohesive function flagged the flake8 too complex error due to all of the branching logic.

this logic follows the same pattern as deploy and destroy